engine: add regression test for batch.ApplyRepr

tbg opened the pull request:
Before the fix, fails with

```
--- FAIL: TestBatchApplyRepr (0.00s)
         batch_test.go:255: old batch represents to:
           "\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x05lost\x00\x06update"
           restored batch to:
           "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
FAIL
```

The fix is not in this commit.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10229)

<!-- Reviewable:end -->

tschottdorf commented:
With this change

```diff
 func (r *rocksDBBatch) Repr() []byte {
-       if r.flushes == 0 {
+       if r.flushes == 0 && false {
                // We've never flushed to C++. Return the mutations only.
                return r.builder.getRepr()
        }
```

We get

```
--- FAIL: TestBatchApplyRepr (0.00s)
       	batch_test.go:255: old batch represents to:
       		"\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x05lost\x00\x06update"
       		restored batch to:
       		"\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x05lost\x00\x06update"
```

which looks suspiciously like it's fixed and the test doesn't realize. I'll see how to clean that up.

tschottdorf commented:
Ok, cleaned up.

petermattis reviewed (approved):
LGTM

Does this affect `applySnapshot`? I think the answer is no, but I'd appreciate a second opinion.

tschottdorf commented:
No, `applySnapshot` wasn't affected because it writes manually after calling `ApplyBatchRepr`. I added an additional test case that shows that `ApplyBatchRepr` followed by `Commit` also lost all the writes (and now doesn't).

petermattis commented on pkg/storage/engine/batch_test.go:
```diff
@@ -228,6 +228,71 @@ func TestBatchRepr(t *testing.T) {
+	e := NewInMem(roachpb.Attributes{}, 1<<20)
+	stopper.AddCloser(e)
+
+	// Failure to represent the absorbed Batch again.
```
I've noticed that you've developed a recent infatuation with the term "absorb".

petermattis commented on pkg/storage/engine/batch_test.go:
```diff
@@ -228,6 +228,71 @@ func TestBatchRepr(t *testing.T) {
+			t.Fatalf("read %q from engine, expected %q", b, val)
+		}
+	}
+
```
Spurious blank line.

tschottdorf replied:
Done. I also named the test `TestApplyBatchRepr` (from `ApplyBatchRepr`, which isn't a test) thanks to `TestUnused`.

tbg committed:
engine: fix add regression test for batch.ApplyRepr

Before the fix, fails with

```
--- FAIL: TestBatchApplyRepr (0.00s)
         batch_test.go:255: old batch represents to:
           "\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x05lost\x00\x06update"
           restored batch to:
           "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
FAIL
```

The fix is avoiding `Repr()` to assume we never flushed to RocksDB (which is
not true after `ApplyRepr()` since that goes directly to RocksDB).
